src/main/java/net/neoforged/neoforge/energy/ComponentEnergyStorage.java

```diff
@@ -0,0 +1,123 @@
+import net.neoforged.neoforge.common.MutableDataComponentHolder;
+
+/**
+ * Variant of {@link EnergyStorage} for use with data components.
+ * <p>
+ * The actual data storage is managed by a data component, and all changes will write back to that component.
+ * <p>
+ * To use this class, register a new {@link DataComponentType} which holds an {@link Integer} for your item.
+ * Then reference that component from your {@link ICapabilityProvider} passed to {@link RegisterCapabilitiesEvent#registerItem} to create an instance of this class.
+ */
+public class ComponentEnergyStorage implements IEnergyStorage {
+    protected final MutableDataComponentHolder parent;
+    protected final DataComponentType<Integer> energyComponent;
+    protected final int capacity;
+    protected final int maxReceive;
+    protected final int maxExtract;
+
+    /**
+     * Creates a new ComponentEnergyStorage with a data component as the backing store for the energy value.
+     * 
+     * @param parent          The parent component holder, such as an {@link ItemStack}
+     * @param energyComponent The data component referencing the stored energy of the item stack
+     * @param capacity        The max capacity of the energy being stored
+     * @param maxReceive      The max per-transfer power input rate
+     * @param maxExtract      The max per-transfer power output rate
+     */
+    public ComponentEnergyStorage(MutableDataComponentHolder parent, DataComponentType<Integer> energyComponent, int capacity, int maxReceive, int maxExtract) {
+        this.parent = parent;
+        this.energyComponent = energyComponent;
+        this.capacity = capacity;
+        this.maxReceive = maxReceive;
+        this.maxExtract = maxExtract;
+    }
+
+    /**
+     * Creates a new ItemEnergyStorage with a unified receive / extract rate.
+     * 
+     * @see ComponentEnergyStorage#ItemEnergyStorage(ItemStack, DataComponentType, int, int, int)
+     */
+    public ComponentEnergyStorage(MutableDataComponentHolder parent, DataComponentType<Integer> energyComponent, int capacity, int maxTransfer) {
+        this(parent, energyComponent, capacity, maxTransfer, maxTransfer);
+    }
+
+    /**
+     * Creates a new ItemEnergyStorage with a transfer rate equivalent to the capacity.
+     * 
+     * @see ComponentEnergyStorage#ItemEnergyStorage(ItemStack, DataComponentType, int, int, int)
+     */
+    public ComponentEnergyStorage(MutableDataComponentHolder parent, DataComponentType<Integer> energyComponent, int capacity) {
+        this(parent, energyComponent, capacity, capacity);
+    }
+
+    @Override
+    public int receiveEnergy(int toReceive, boolean simulate) {
+        if (!canReceive() || toReceive <= 0) {
+            return 0;
+        }
+
+        int energy = this.getEnergyStored();
+        int energyReceived = Mth.clamp(this.capacity - energy, 0, Math.min(this.maxReceive, toReceive));
+        if (!simulate) {
+            this.setEnergy(energy + energyReceived);
```

Comment:
If energyReceived is zero, because it is already full we want to skip bothering with the component mutation via setEnergy. And likewise for extractEnergy if it is already empty